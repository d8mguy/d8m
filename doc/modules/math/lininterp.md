## Lininterp

Some linear interpolation code. While linear interpolation itself is a simple and well defined thing, there are quite a few ways to present the specification, deal with boundary conditions, handle slope calculations, etc.

This module is a stub, which provides one way of doing it; my intention is to provide a few variants in the fullness of time. For example, a not yet written algorithm could handle interpolants in any order by storing all slopes and using binary search for indices. 

 

#### type interpFSM

    [interpFSM: xlst, ylst]
where `xlst, ylst: list(float)`. Panic unless `xlst` and `ylst` have same length and that length is > 1.

`InterpFSM` holds state to generate a sequence of float interpolants from a non-decreasing series of `x:float` query values. Only calculates slopes as needed for the presented interpolant `x` values, so may be faster than some algorithms. The `exists` method catches values outside the x range of the interpolable data points; this allows the client to implement its preferred policy. 

    val interpFSM = extend tuple(curinx: integer, xs, ys: list(float), slope:float, noslope:boolean) where {
        method reset = \mod() -> nothing {...}
        method exists = \(x:float) -> boolean {...}
        method next = \mod(x:float) -> float {...}
    }

Method `reset`: Set internal state to same as newly constructed.

Method `exists`: true if self can supply an interpolant at x.

Method `next`: return the interpolant for x and update state, assuming the next `x` given will be >= this one. Assume `self.exists(x)`.

