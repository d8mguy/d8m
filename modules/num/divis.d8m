// some code around divisibility ftns


export gcd, gcdlist, lcm, lcmlist

// gcd of two non-neg integers
val gcd = \(a,b:integer) -> integer {
    if(a == 0) b else gcd(b % a, a)
}

// This could be named gcd in d8m, since it's TC distinct from the one above. But for some reason I prefer not to.
val gcdlist = \imp(lst:list(integer)) {
    if(lst.count == 0) 0        // this is really an error
    else {
        var soFar = lst[0]
        each(elt^lst.tail) {
            soFar = gcd(soFar, elt)
            if(soFar == 1) break
        }
        soFar
    }
}

// A not particularly efficient way to calculate the LCM
val lcm = \(a,b:integer) { a*b/gcd(a,b) }
val lcmlist = \imp(lst:list(integer)) {
    if(lst.count == 0) 0        // this is really an error
    else reduce(lst.tail, \(x,y:integer) { lcm(x,y)}, lst[0])
}
