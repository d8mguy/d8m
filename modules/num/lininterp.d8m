// some linear interpolation code
// My intention is to provide a few variants in the fullness of time. For example, interpUnordered (not written)
// could handle interpolants in any order by storing all slopes and using binsearch for indices.

export interpFSM

// InterpFSM holds state to generate a sequence of float interpolants from a non-decreasing series of x:float query values.
// There are three methods: reset, exists and next. Reset restarts the sequence. Exists returns whether a proposed x value
// can be interpolated in the given xs, ys series. This approach allows users to have their own policy for values outside
// the range. Assuming exists(x) is true, next(x) returns the interpolant of x AND updates self to assume future requests will be >= x.
// It does a linear search for the indices of values in xs, which should be strictly increasing. This many not be optimal
// if xs is long and only a few points are sampled.
// I think the performance tradeoff on slope calculation is good: if few points are sampled, then only needed slopes are
// calculated, but if an interval is used more than once, its slope is only calculated once. This makes sense only under
// the assumption that we step left to right through a sequence of interpolants.
// Panic unless xs and ys have same length and that length is > 1.
val interpFSM = extend tuple(curinx: integer, xs, ys: list(float), slope:float, noslope:boolean) where {
    assert xs.count != 0
    method lithook = \mod(xlst, ylst: list(float)) {
        if(xlst.count == 0 || ylst.count != xlst.count) exit("interpFSM")
        xs = xlst
        ys = ylst
        curinx = 0
        noslope = true
        //println("creating fsm, xs=", xs.{to_s(this,2)}, "; ys=", ys.{to_s(this,2)})
    }

    // Set internal state to same as newly constructed.
    method reset = \mod() -> nothing { curinx = 0; noslope = true }

    // true if self can supply an interpolant at x
    method exists = \(x:float) {
        //println("exists:", x.to_s(1), curinx)
        x >= xs[curinx] && x <= xs.last
    }

    // Return the interpolant for x and update state, assuming the next x given will be >= this one.
    // Assume self.exists(x).
    method next = \mod(x:float) -> float {
        //println("next", to_s(x, 2))
        if(x == xs.last) ys[ys.count - 1]
        else {
            while(xs[curinx+1] <= x) { curinx += 1; noslope = true }
            if(noslope) { slope = (ys[curinx+1] - ys[curinx]) / (xs[curinx+1] - xs[curinx]); noslope = false }
            //println(curinx, to_s(xs[curinx], 1), to_s(ys[curinx+1], 1), to_s(slope, 3), to_s(ys[curinx] + slope * (x - xs[curinx]), 3))
            ys[curinx] + slope * (x - xs[curinx])
        }
    }
}
