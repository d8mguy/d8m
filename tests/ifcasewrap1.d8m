// This test isn't about lininterp but to ensure that the case clauses get wrapped in stmts so that backend doesn't move
// code out of the case stmt.
import "num/lininterp"
import go "math/rand"

val foo = \imp() {
    val randx = rand.New(rand.NewSource(1))
    var compares: list(float) = []
    val xvals0 = (1..10).{ 4.0 + 1.5*this }
    val xvals1 = (1..10).{ 3.0 + 1.9*this }
    val cvals0 = (1..10).{ 15.0 + 0.6*this }
    val cvals1 = (1..10).{ 20.0 + 1.2*this }
    var fsm = [lininterp.interpFSM: xvals0, cvals0]
    val cmprOpts = 1+randx.Intn(3)
    case {
        cmprOpts == 1 => compares = [x0~xvals0, y0~cvals0].xp.{ fsm.exists(x0) ? y0/fsm.next(x0) : 0.0 }
        cmprOpts == 2 => compares = [x0~xvals0, y0~cvals0].xp.{ fsm.exists(x0) ? y0 - fsm.next(x0) : -1.0 }
        cmprOpts == 3 => compares = [x0~xvals1, y0~cvals1].xp.{ fsm.exists(x0) ? y0/fsm.next(x0) : 0.0 }
        cmprOpts == 4 => compares = [x0~xvals1, y0~cvals1].xp.{ fsm.exists(x0) ? y0 - fsm.next(x0) : -1.0 }
    }
    compares
}
foo()
