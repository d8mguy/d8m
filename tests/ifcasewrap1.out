[0.000000, 1.389578, 1.398104, 1.405896, 1.413043, 1.419624, 1.425703, 1.431335, 0.000000, 0.000000]
